sky/spot/controller.py

```diff
@@ -309,14 +311,25 @@ def start(job_id, task_yaml, retry_until_up):
     except exceptions.SpotUserCancelledError:
         logger.info(f'Cancelling spot job {job_id}...')
         if controller_process is not None:
-            logger.info('Sending SIGTERM to controller process '
+            logger.info('Killing controller process '
                         f'{controller_process.pid}')
-            # This will raise KeyboardInterrupt in the task.
-            # Using SIGTERM instead of SIGINT, as the SIGINT is weirdly ignored
-            # by the controller process when it is started inside a ray job.
-            controller_process.terminate()
-    if controller_process is not None:
-        controller_process.join()
+            spot_state.set_cancelled(job_id)
+    finally:
+        while controller_process is not None and controller_process.is_alive():
+            # NOTE: it is ok to kill or join a killed process.
+            controller_process.kill()
+            # Kill the children processes launched by log_lib.run_with_log.
+            subprocess_utils.kill_children_processes()
```

Comment:
it will work. it was never `daemon` process in the original PR, so it does not make sense that it will not work here

Reply:
Well, I think previously, the children processes are the spawn with `subprocess.Popen` directly from the current process, instead of having an additional `multiprocessing.Process` in between, i.e. previously the `kill_children_processes` is directly called within the `controller_processs` instead of the outer caller. I am not sure if that will make any difference.

Reply:
From my observation it works, also as the note suggests it supports general *nix process:

```
        """Return the children of this process as a list of Process
        instances, pre-emptively checking whether PID has been reused.
        If *recursive* is True return all the parent descendants.

        Example (A == this process):

         A ─┐
            │
            ├─ B (child) ─┐
            │             └─ X (grandchild) ─┐
            │                                └─ Y (great grandchild)
            ├─ C (child)
            └─ D (child)

        >>> import psutil
        >>> p = psutil.Process()
        >>> p.children()
        B, C, D
        >>> p.children(recursive=True)
        B, X, Y, C, D

        Note that in the example above if process X disappears
        process Y won't be listed as the reference to process A
        is lost.
        """
```